### ISSUE TITLE
Retrying the insert on issue fails, and masks the actual error

### ISSUE THREAD

**Issue by guillep2k** [go-gitea/gitea#8302]
- Gitea version (or commit ref): c6fb7fe27c16c4e43d4d8dbe4d2ff4b3c4c52a29

## Description

The changes introduced in  #7898 simply don't work, because if `xorm.Session.Insert()` returns any errors, the session becomes invalidated, so subsequent attempts fail with:

```
newIssue: pq: current transaction is aborted, commands ignored until end of transaction block
```

https://github.com/go-gitea/gitea/blob/c6fb7fe27c16c4e43d4d8dbe4d2ff4b3c4c52a29/models/issue.go#L1108-L1122

I'm sorry, after so many attempts to solve the potential duplicate `index`, we're back to square zero.

@lunny how would you like to handle this?

**Comment by lunny** [comment]
The retry should be done after the transaction end. It should be out of `newIssue` but we should return the reason of failed. So that we could know if we need retry.

**Comment by guillep2k** [comment]
> 
> 
> The retry should be done after the transaction end. It should be out of `newIssue` but we should return the reason of failed. So that we could know if we need retry.

I agree, but as I've [said here](https://github.com/go-gitea/gitea/pull/7898#issue-308264711), detecting the _correct_ error is very difficult.

**Comment by lunny** [comment]
If the error is insert error we could retry it, I think.

**Comment by guillep2k** [comment]
> 
> 
> If the error is insert error we could retry it, I think.

Oh, you mean _any_ errors, as long as they come from the insert? That's much easier. I'll work on that. I'll asume that #8304 will be merged and I'll base on it.

**Comment by lunny** [comment]
Should be fixed by #8307

**Comment by guillep2k** [comment]
Agree
